### PR TITLE
Add a NodeCondition "NetworkUnavailable" to prevent scheduling onto a node until the routes have been created 

### PR DESCRIPTION
**Pull request by wojtek-t**
This is new version of #26267 (based on top of that one).

The new workflow is:
- we have an "NetworkNotReady" condition
- Kubelet when it creates a node, it sets it to "true"
- RouteController will set it to "false" when the route is created
- Scheduler is scheduling only on nodes that doesn't have "NetworkNotReady ==true" condition

**Release Note**

``` release-note
In GCE/GKE route has to be created for a node, in order to make it schedulable.
As long as route is not created, a node will have "NetworkUnavailable" condition set, which will prevent scheduler from scheduling anything on that node.
```

@gmarek @bgrant0607 @zmerlynn @cjcullen @derekwaynecarr @danwinship @dcbw @lavalamp @vishh

### CODE REVIEW

**Comment by wojtek-t** [comment]
There is currently a problem with this PR - kubelet is overwriting node status written by route controller.

**Comment by wojtek-t** [comment]
Actually something different is happening here. Investigating.

**Comment by wojtek-t** [comment]
OK - debugged and fixed. Hopefully tests will pass now.

**Comment by derekwaynecarr** [comment]
This looks better to me.

@kubernetes/rh-networking - please take a pass.

File: pkg/api/types.go [hunk 1]
```diff
@@ -2009,6 +2009,8 @@ const (
 	NodeOutOfDisk NodeConditionType = "OutOfDisk"
 	// NodeMemoryPressure means the kubelet is under pressure due to insufficient available memory.
 	NodeMemoryPressure NodeConditionType = "MemoryPressure"
+	// NodeNetworkingNotReady means that network for the node is not correctly configured.
+	NodeNetworkingNotReady NodeConditionType = "NetworkingNotReady"
```
Comment (bgrant0607) [inline pkg/api/types.go]: I prefer "NetworkUnavailable".

File: pkg/kubelet/kubelet.go [hunk 2]
```diff
@@ -1078,6 +1089,14 @@ func (kl *Kubelet) initialNodeStatus() (*api.Node, error) {
 	return node, nil
 }
 
+func (kl *Kubelet) providerRequiresNetworkingConfiguration() bool {
+	if kl.cloud == nil || kl.flannelExperimentalOverlay {
+		return false
+	}
+	// TODO: Should we include any other providers here?
+	return kl.cloud.ProviderName() == gce.ProviderName
```
Comment (bgrant0607) [inline pkg/kubelet/kubelet.go]: This belongs in the cloudprovider package. We don't check provider names outside that package.

File: pkg/kubelet/kubelet.go [hunk 3]
```diff
@@ -1013,6 +1014,16 @@ func (kl *Kubelet) initialNodeStatus() (*api.Node, error) {
 			Unschedulable: !kl.registerSchedulable,
 		},
 	}
+	// Initially, set NetworkCondition to false.
```
Comment (bgrant0607) [inline pkg/kubelet/kubelet.go]: Update the name of the condition.

File: pkg/controller/route/routecontroller.go [hunk 4]
```diff
@@ -155,6 +162,63 @@ func (rc *RouteController) reconcile(nodes []api.Node, routes []*cloudprovider.R
...
+func (rc *RouteController) updateNetworkingCondition(nodeName string, routeCreated bool) error {
+	var err error
+	for i := 0; i < updateNodeStatusMaxRetries; i++ {
+		node, err := rc.kubeClient.Core().Nodes().Get(nodeName)
+		if err != nil {
+			glog.Errorf("Error geting node: %v", err)
+			continue
+		}
+		updateNetworkingCondition(node, routeCreated)
+		if _, err = rc.kubeClient.Core().Nodes().UpdateStatus(node); err == nil {
```
Comment (bgrant0607) [inline pkg/controller/route/routecontroller.go]: https://github.com/kubernetes/node-problem-detector/issues/9

Comment (bgrant0607) [inline pkg/controller/route/routecontroller.go]: https://github.com/kubernetes/kubernetes/pull/26381

File: pkg/api/types.go (hunk 1 above)
Comment (wojtek-t) [inline pkg/api/types.go]: Changed.

File: pkg/controller/route/routecontroller.go (hunk 4 above)
Comment (wojtek-t) [inline pkg/controller/route/routecontroller.go]: Added a TODO to fix that.

File: pkg/kubelet/kubelet.go (hunk 3 above)
Comment (wojtek-t) [inline pkg/kubelet/kubelet.go]: Done

File: pkg/kubelet/kubelet.go (hunk 2 above)
Comment (wojtek-t) [inline pkg/kubelet/kubelet.go]: Changed to use only standard things.

**Comment by wojtek-t** [comment]
@bgrant0607 - thanks a lot for quick review.

Comments applied. PTAL

BTW - do we need want a release note for this?

**Comment by bgrant0607** [comment]
LGTM

**Comment by wojtek-t** [comment]
Thanks @bgrant0607 !

**Comment by vishh** [comment]
LGTM

**Comment by erictune** [comment]
@wojtek-t  Does this PR require action by the user when upgrading from 1.2.x to 1.3.0?  (Think about non-developer users.) If so, please edit your first comment to have  a release-note block, like in https://github.com/kubernetes/kubernetes/pull/28132.  If it is just an optional feature, please change the label to just release-note.  If it is not a complete feature by itself, then apply "release-note-none" label instead.

**Pull request merged**